### PR TITLE
ADBDEV-5829 Refactor ListConstToStr function

### PR DESCRIPTION
**Pull request by bimboterminator1**
The original implementation of the list_const_to_str (ListConstToStr) function
was intended to extract and format values from array constants into a string
buffer, supporting data types like int2[], int4[], int8[], and text[]. It
checked for null constants, logged exceptions, extracted the array, processed
each supported type with specific code blocks to deconstruct the array,
converted each element to a string, and appended the formatted data to a buffer.
However, the approach led to code duplication, maintenance challenges, and
reduced readability due to scattered, repetitive logic.

The refactored function consolidated common logic into reusable, unified steps.
It retrieves type information and deconstructs arrays using a single,
generalized procedure, eliminating repetitive code. This streamlined array
processing is supported by getTypeOutputInfo, which extracts output function
from the catalog, facilitating consistent conversion across all data types
with OidOutputFunctionCall.

Additionally, the refactored function simplifies its logic by centralizing the
handling of array elements and reducing the switch case complexity. This
enhances readability and maintainability, making it easier to understand and
modify. Adding new array types now involves minimal changes, creating a more
elegant, less error-prone codebase.

----
### How to test with jdbc:
With external table:
1. create jdbc profile in $PXF_BASE/conf/pxf-profiles.xml
<details>
  <summary>Ex</summary>

 ```

<?xml version='1.0' encoding='UTF-8'?>
<profiles>
  <profile>
    <name>JDBC</name>
    <description>jdbc</description>
    <plugins>
      <fragmenter>org.greenplum.pxf.plugins.jdbc.JdbcPartitionFragmenter</fragmenter>
      <accessor>org.greenplum.pxf.plugins.jdbc.JdbcAccessor</accessor>
      <resolver>org.greenplum.pxf.plugins.jdbc.JdbcResolver</resolver>
    </plugins>
    </profile>
</profiles>

``` 
 
</details>
2. create server configuration in $PXF_BASE/servers/postgresql/jdbc-site.xml
<details>
  <summary>Ex</summary>

 ```
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <property>
        <name>jdbc.driver</name>
        <value>org.postgresql.Driver</value>
        <description>Class name of the JDBC driver (e.g. org.postgresql.Driver)</description>
    </property>
    <property>
        <name>jdbc.url</name>
        <value>jdbc:postgresql://localhost:5432/postgres</value>
        <description>The URL that the JDBC driver can use to connect to the database (e.g. jdbc:postgresql://localhost/postgres)</description>
    </property>
    <property>
        <name>jdbc.user</name>
        <value>bimbo</value>
        <description>User name for connecting to the database (e.g. postgres)</description>
    </property>
</configuration>

``` 
</details>

3.
```
create extension pxf;
CREATE readable EXTERNAL TABLE test_int2(i int2)
     LOCATION ('pxf://public.test_int2?PROFILE=JDBC&SERVER=postgresql')
          FORMAT 'CUSTOM' (FORMATTER='pxfwritable_import');

select * from test_int2 where i in (1::int2, 3::int2);
select * from test_int4 where i in (1, 3);
select * from test_int8 where i in (1::int8, 3::int8);
select * from test_text where i in ('asdasdas'::text, 'tsadasd'::text);
```
With fdw:
1. Create configuration:
<details>
  <summary>Ex</summary>

 ```
CREATE SERVER postgresql_server
     FOREIGN DATA WRAPPER jdbc_pxf_fdw
          OPTIONS ( 
         jdbc_driver 'org.postgresql.Driver', 
         db_url 'jdbc:postgresql://localhost:5432/postgres',
          batch_size '10000',
         fetch_size '2000'
     );
     
     
 CREATE USER MAPPING FOR my_user
     SERVER postgresql_server
     OPTIONS ( user 'pg_user');

``` 
 
</details>

2. The same queries:
```
           CREATE FOREIGN TABLE test_text_fdw(t text)
          SERVER postgresql_server
          OPTIONS ( resource 'public.test_text');

select * from test_text_fdw where i in ('asdasdas'::text, 'tsadasd'::text);
```
And then just look at pxf-service.log or [here](https://github.com/arenadata/pxf/blob/103d3008fa6a1f02ec4b2bc47e9ffe242822558b/external-table/src/pxfheaders.c#L211)

### CODE REVIEW

File: external-table/src/pxffilters.c [hunk 1]
```diff
@@ -1217,81 +1222,40 @@ list_const_to_str(Const *constval, StringInfo buf)
 
 	interm_buf = makeStringInfo();
 
+	/*
```
Comment (whitehawk) [inline external-table/src/pxffilters.c]: I think it is better to move the entire block

```
	arr = DatumGetArrayTypeP(constval->constvalue);

	interm_buf = makeStringInfo();

	/*
	 * Get necessary data for deconstruct_array() and output function
	 * from the catalog.
	 */
	get_typlenbyvalalign(ARR_ELEMTYPE(arr), &elmlen, &elmbyval, &elmalign);
	deconstruct_array(arr,
					  ARR_ELEMTYPE(arr),
					  elmlen,
					  elmbyval,
					  elmalign,
					  &dats,
					  NULL,
					  &len);

	getTypeOutputInfo(ARR_ELEMTYPE(arr), &typoutput, &typIsVarlena);
```

inside the `switch-case` right before the loop. There is no reason to make all these steps if we end up at `elog(ERROR,...`. Let's do them only for the types we can handle here.

Or we can just move the type check to the beginning of the function...

File: external-table/src/pxffilters.c [hunk 2]
```diff
@@ -1301,7 +1265,6 @@ list_const_to_str(Const *constval, StringInfo buf)
 				 "internal error in pxffilters.c:list_const_to_str. "
 				 "Using unsupported data type (%d)",
 				 constval->consttype);
-
```
Comment (whitehawk) [inline external-table/src/pxffilters.c:1304]: why the empty line was removed? maybe better to revert this change to reduce delta?

File: external-table/src/pxffilters.c [hunk 3]
```diff
@@ -1220,78 +1220,42 @@ list_const_to_str(Const *constval, StringInfo buf)
...
+				Oid			typoutput;
+				bool		typIsVarlena;
+				int16		elmlen;
+				bool		elmbyval;
+				char		elmalign;
+				/*
+				 * Get necessary data for deconstruct_array() and output function
+				 * from the catalog.
+				 */
+				get_typlenbyvalalign(ARR_ELEMTYPE(arr), &elmlen, &elmbyval, &elmalign);
```
Comment (whitehawk) [inline external-table/src/pxffilters.c:1235]: Why not to move 
```
	arr = DatumGetArrayTypeP(constval->constvalue);
	interm_buf = makeStringInfo();
```

and 
```
pfree(interm_buf->data);
```
here as well?

Comment (bimboterminator1) [reply external-table/src/pxffilters.c:1235]: This is suitable in current supported type range. I'll move this code in this patch. However, i'll probably reconsider this function again while working at ADBDEV-5772, and i'll probably change the code.

Comment (whitehawk) [reply external-table/src/pxffilters.c:1235]: ok

**Review by whitehawk — APPROVED** [review]
